File: Behavioral/Lab19/Typescript/src/Directors/luxury_estate_director.ts
import { EstateBuilder } from "../Builders/estate_builder";
import { EstateDirector } from "./estate_director";
import { RealEstate } from "../Entities/real_estate";
import { HouseBuilder } from "../Builders/house_builder";
import { ApartmentBuilder } from "../Builders/apartment_builder";
import { RoomBuilder } from "../Builders/room_builder";

export class LuxuryEstateDirector extends EstateDirector {

    constructor(estateBuilder: EstateBuilder) {
        super(estateBuilder);
    }
    public BuildDefault():RealEstate{
        const type = this._estateBuilder.constructor.name.replace("Builder", "").toLowerCase();
        this._estateBuilder.reset();
        this._estateBuilder
            .setAddress("")
            .setDescription(`Luxury ${type}`)
            .setPrice(100000)
            .setStatus("New")
            .setFurnished()
            .setAirConditioning();

        switch (type) {
            case "house":
                const h = this._estateBuilder as HouseBuilder;
                return h
                    .setPool()
                    .setGym()
                    .setGarden()
                    .setHeatingType("local")
                    .getEstate();
            case "apartment":
                const a = this._estateBuilder as ApartmentBuilder;
                return a
                    .setRoomsNumber(5)
                    .setHeatingType("central")
                    .setElevator()
                    .setParking()
                    .getEstate();
            case "room":
                const r = this._estateBuilder as RoomBuilder;
                return r
                    .setNeighbors(true)
                    .setHeatingType("central")
                    .getEstate();
            default:
                throw new Error("Wrong type of estate.");
        }
    }
}

File: Behavioral/Lab22/C#/Directors/LuxuryEstateDirector.cs
using System;
using System.Collections.Generic;
using System.Text;
using C_.Builders;
using C_.Entities;

namespace C_.Directors
{
    public class LuxuryEstateDirector : EstateDirector
    {
        public LuxuryEstateDirector(EstateBuilder estateBuilder):base(estateBuilder){}
        public override RealEstate BuildDefault()
        {
            var type = _estateBuilder.GetType().Name.Replace("Builder", "");
            _estateBuilder.Reset();
            _estateBuilder
                .SetAddress("")
                .SetDescription("Luxury " + type.ToLower())
                .SetPrice(100000)
                .SetStatus("New")
                .SetFurnished()
                .SetAirConditioning();
            
            switch(type.ToLower()){
                case "house": 
                    var h = (HouseBuilder)_estateBuilder;
                    return h
                        .SetPool()
                        .SetGym()
                        .SetGarden()
                        .SetHeatingType("local")
                        .GetEstate();
                case "apartment":
                    var a = (ApartmentBuilder)_estateBuilder;
                    return a
                            .SetRoomsNumber(5)
                            .SetHeatingType("central")
                            .SetElevator()
                            .SetParking()
                            .GetEstate();
                case "room":
                    var r = (RoomBuilder)_estateBuilder;
                        return r
                            .SetNeighbors(true)
                            .SetHeatingType("central")
                            .GetEstate();
                default: 
                    throw new Exception("Wrong type of estate.");
            }
        }
    }
}
